text/0000-float-semantics.md

```diff
@@ -0,0 +1,314 @@
+
+# Motivation
+[motivation]: #motivation
+
+We have a plethora of open issues that boil down to "is this sequence of float operations allowed to produce the given result".
+This is caused by a combination of surprising effects introduced by LLVM optimizations, bugs in MIR and LLVM optimizations, and bugs (or at least non-conformance) in certain targets. See [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) for a collection of issues.
+
+It's time to stop leaving our users in the dark about what actually is and is not guaranteed.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Primitive operations on floating-point types generally produce results that exactly match IEEE 754-2008:
+if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will not be able to observe any non-determinism and behave according to the IEEE specification.
+
+If you *do* use these operations on NaNs, then the exact behavior you see can depend on compiler version, compiler flags, target architecture, and it can even be non-deterministic (i.e., running the same operation on the same inputs twice can yield different results).
+The results produced in these cases do *not* always conform to the IEEE specification.
+See the reference section for what exactly is guaranteed.
+
+When a floating-point value is just passed around, its contents (including the bits of a NaN) do *not* change.
+
+When you use a floating-point operation in [`const` context](https://doc.rust-lang.org/reference/const_eval.html#const-context), the same specification applies: NaN bit patterns are non-deterministic.
+In particular, the bit pattern produced at compile-time can differ from the bit pattern produced by the same operation at run-time.
+
+Certain targets unfortunately are known to not implement these semantics precisely (see [below](#target-specific-problems)).
+The [platform support page](https://doc.rust-lang.org/rustc/platform-support.html) will list those caveats.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC specifies the behavior of `+`, `-` (unary and binary), `*`, `/`, `%`, `abs`, `copysign`, `mul_add`, `sqrt`, `as`-casts that involve floating-point types, and all comparison operations on floating-point types.
+Here, "floating-point types" are `f32` and `f64` and all similar types that might be added in the future such as `f16`, `f128`.
+Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven [except for float-to-int casts, which round towards zero] and default exception handling without traps, without abruptUnderflow/flush-to-zero).
```

Comment:
```suggestion
Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven [except for float-to-int casts, which round towards zero](...) and default exception handling without traps, without abruptUnderflow/flush-to-zero).
```

The link seems to miss a URL.

Reply:
No, these are just nested parentheses.

Reply:
Oh, I see. Thanks for the clarification.